src/AndersonAcceleration.f90

```diff
@@ -196,59 +143,217 @@ SUBROUTINE init_AndersonAccelerationType(solver,MPIEnv,Params)
 !> @brief Clears the Anderson Acceleration Type
 !> @param solver The solver to act on
 !>
-!> This routine clears the data spaces
-!>
 SUBROUTINE clear_AndersonAccelerationType(solver)
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
 
-  NULLIFY(solver%MPIparallelEnv)
-  IF(solver%OMPparallelEnv%isInit()) CALL solver%OMPparallelEnv%clear
-  solver%iter=0
-  solver%n=-1
-  solver%depth=-1
-  solver%beta=0.0_SRK
-  IF(ASSOCIATED(solver%X)) THEN
-    CALL solver%X%clear()
-    DEALLOCATE(solver%X)
-    NULLIFY(solver%X)
+  IF(solver%isInit) THEN
+    NULLIFY(solver%MPIparallelEnv)
+    solver%s=0
+    solver%N=-1
+    solver%depth=-1
+    solver%start=1
+    solver%depth=-1
+    solver%beta=0.0_SRK
+    DEALLOCATE(solver%x)
+    DEALLOCATE(solver%Gx)
+    DEALLOCATE(solver%r)
+    DEALLOCATE(solver%A)
+    DEALLOCATE(solver%alpha)
+    DEALLOCATE(solver%b)
+    solver%isInit=.FALSE.
   ENDIF
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  CALL Anderson_Destroy(solver%id)
-#endif
-  solver%isInit=.FALSE.
+
 ENDSUBROUTINE clear_AndersonAccelerationType
 !
 !-------------------------------------------------------------------------------
-!> @brief Single step of the Anderson acceleration
-!> @param solver The anderson solver to act on
+!> @brief Single step of Anderson acceleration acting on solution vectors
+!> @param solver Anderson solver to take step with
+!> @param New_Gx new solution iterate
 !>
 !> This routine takes a single anderson acceleration step
 !>
-SUBROUTINE step_AndersonAccelerationType(solver)
+FUNCTION step_AndersonAccelerationType(solver,New_Gx) RESULT(x_tmp)
```

Comment:
These need to be vector classes so you can maintain your generic interface.

Reply:
the more I think about this, I'm a little uneasy about this being a function.  What happens if the return value isn't the same type as what gets passed in.  
```
REAL :: x(5)
x = foo(3) ! tries to return size (8) array... gets more complicated for classes.
```